### PR TITLE
Add loading spinner for connection tree element refresh

### PR DESCRIPTION
**Pull request by smartguest**
PR adds in a spinner to the element that is refreshed (unless it is an item with no children), once the item is immediately refreshed the spinner will show for a second later to indicate that a refresh happened (as it's usually instantaneous). Child elements do not have a spinner shown ATM, but it is not necessary because items will only appear once they are loaded.

This PR fixes #https://github.com/microsoft/azuredatastudio/issues/17458

GIF of fix in action:

RefreshServerExplorerUpdated

### CODE REVIEW

File: src/vs/base/parts/tree/browser/treeModel.ts [hunk 1]
```diff
@@ -969,7 +972,17 @@ export class TreeModel {
 		let eventData: IRefreshEvent = { item: item, recursive: recursive };
 		this._onRefresh.fire(eventData);
 		return item.refresh(recursive).then(() => {
```
Comment (Charles-Gagnon) [inline src/vs/base/parts/tree/browser/treeModel.ts:974]: While you're here let's make this async to simplify it

File: src/vs/base/parts/tree/browser/treeModel.ts [hunk 2]
```diff
@@ -969,7 +972,17 @@ export class TreeModel {
...
+			/**
+			 * Refresh loading status for item will be instantly lost upon refresh of item self
+			 * (usually before the spinner shows). we need to temporarily bring it back upon
+			 * actual refresh completion to let the user know the refresh happened with a short
+			 * appearance of the loading spinner.
+			 */
+			this.addTraits('loading', [element]);
 			this._onDidRefresh.fire(eventData);
+			setTimeout(function () {
```
Comment (Charles-Gagnon) [inline src/vs/base/parts/tree/browser/treeModel.ts:983]: If you use a lambda function (`() => { ... }`) then you don't need a separate var to capture this, it will do it automatically. (i.e. you can remove the self var)

**Review by Charles-Gagnon — COMMENTED** [review]
Please add a gif showing the new behavior

File: src/vs/base/parts/tree/browser/treeModel.ts [hunk 3]
```diff
@@ -960,6 +960,9 @@ export class TreeModel {
 	}
 
 	public refresh(element: any = null, recursive: boolean = true): Promise<any> {
+		let self = this;
+		// Add loading spinner for item during refresh process.
+		this.addTraits('loading', [element]);
```
Comment (Charles-Gagnon) [inline src/vs/base/parts/tree/browser/treeModel.ts:965]: So any changes in the VS folder should always have a SQL CARBON EDIT. But in this case this stuff is a super old version of the tree that VS removed a long time back (https://github.com/microsoft/vscode/commit/f28dbf7cfc24c876dae2b006d3b3b381b7b722f9) so this is fine, although I think I'll go and actually move this into the sql folder then to reduce the "extra" stuff we have in the VS folder. 

(this is more of a general comment than anything you need to do anything about)

File: src/vs/base/parts/tree/browser/treeModel.ts (hunk 1 above)
Comment (smartguest) [reply src/vs/base/parts/tree/browser/treeModel.ts:974]: Do you mean the inner function or the refresh function itself?

File: src/vs/base/parts/tree/browser/treeModel.ts (hunk 3 above)
Comment (smartguest) [reply src/vs/base/parts/tree/browser/treeModel.ts:965]: Didn't realize it was in the vs folder until now.

**Comment by smartguest** [comment]
RefreshServerExplorer
GIF of the refresh spinner in action (items without children won't show anything, such as column items in a table)

**Comment by cheenamalhotra** [comment]
@smartguest I ran this locally, and it doesn't seem to refresh a "Closed" connection, can you try the same and confirm with the gif in case I missed something?

File: src/vs/base/parts/tree/browser/treeModel.ts [hunk 4]
```diff
@@ -968,9 +970,18 @@ export class TreeModel {
 
 		let eventData: IRefreshEvent = { item: item, recursive: recursive };
 		this._onRefresh.fire(eventData);
-		return item.refresh(recursive).then(() => {
-			this._onDidRefresh.fire(eventData);
-		});
+		await item.refresh(recursive);
+		/**
+		 * Refresh loading status for item will be instantly lost upon refresh of item self
```
Comment (Charles-Gagnon) [inline src/vs/base/parts/tree/browser/treeModel.ts:975]: So am I correct in my understanding that this is how it currently works? (especially the 1sec delay time)

1. We add the loading trait to the item before starting the refresh
2. The item "finishes" the refresh call immediately so the loading spinner disappears (but the actual refresh call is still happening in the background because of how OE does refreshes)
3. You add the loading trait back in again because it's not actually "done". But this loading spinner is hardcoded to always be around for 1sec - no matter how long the refresh takes
4. Eventually it finishes the refresh and actually updates the node

File: src/vs/base/parts/tree/browser/treeModel.ts [hunk 5]
```diff
@@ -959,7 +959,9 @@ export class TreeModel {
 		return this.input ? this.input.getElement() : null;
 	}
 
-	public refresh(element: any = null, recursive: boolean = true): Promise<any> {
+	public async refresh(element: any = null, recursive: boolean = true): Promise<any> {
+		// Add loading spinner for item during refresh process.
+		this.addTraits('loading', [element]);
```
Comment (Charles-Gagnon) [inline src/vs/base/parts/tree/browser/treeModel.ts:964]: Shouldn't this go below the `if (!item)` check on 967? If we don't have an item there's nothing we should be adding the trait to.

File: src/vs/base/parts/tree/browser/treeModel.ts (hunk 4 above)
Comment (smartguest) [reply src/vs/base/parts/tree/browser/treeModel.ts:975]: Yeah that is correct, it isn't accurate to actual loading status, but the addTrait call requires an item to update, and I'm not sure where in the code does the actual refresh finishes (I tried to dig through, but I could never find it while debugging as there were a ton of recursion)

Comment (Charles-Gagnon) [reply src/vs/base/parts/tree/browser/treeModel.ts:976]: Ok, well that doesn't really sound like a good fix to me. It's going to pretty much always be wrong - and refreshes can take a long time so if they take >1sec then it's just going to be confusing ("the loading spinner went away but the nodes were the same!")

If we're going to fix this it should be done correctly. (keep in mind that other things like the connection browse tree in the connection widget use this same tree structure so you probably shouldn't be inserting any connection-specific logic into the tree)

**Comment by smartguest** [comment]
The test failure in this PR is not caused by this change specifically, when I ran the unit test on the main branch, it had the same unexpected result as well.

**Comment by Charles-Gagnon** [comment]
Where are you seeing that failing in a build from main? The daily builds are passing so it seems to indicate that this PR is causing the failure (and the error is from treeModel which you've changed here so it seems even more likely that something in your PR is breaking that test)

**Comment by smartguest** [comment]
The test failure that is happening here is not caused by this PR specifically, it reproduces on the main branch when I run it on my side.

**Comment by smartguest** [comment]
It happens on my local branch when I run the unit test on it's own, I have no idea why it is not working ATM

**Comment by smartguest** [comment]
Just made it so that upon refresh completion of the item's children, and/or the item itself in the case of it being collapsed, it will show the loading spinner then, this one is more accurate in tracking the loading status of the item (without going into the provider itself).

File: src/sql/base/parts/tree/browser/treeModel.ts [hunk 6]
```diff
@@ -525,12 +533,34 @@ export class Item {
...
+		 * actual refresh completion to let the user know the refresh happened with a short
+		 * appearance of the loading spinner.
+		 */
+		Item.originalItem?.addTrait('loading');
+		this.refreshLoadTimeout = setTimeout(() => {
+			Item.originalItem?.removeTrait('loading');
+			// Reset original item for next refresh.
+			Item.originalItem = undefined;
+			this.refreshLoadTimeout = undefined;
+		}, 1000);
```
Comment (Charles-Gagnon) [inline src/sql/base/parts/tree/browser/treeModel.ts:561]: So you're still doing that "show it for 1sec regardless of how long it actually takes to fully refresh"? That still doesn't seem like the solution we want here - if we're going to fix this it seems like something we should actually fix completely so it shows the loading spinner while it's refreshing and not any more/less. 

(We don't want to be in a situation where the refresh takes longer and the spinner disappears but then the refresh isn't quite finished. Same but to a slightly lesser extent for refresh being quicker than the timeout)

Comment (smartguest) [reply src/sql/base/parts/tree/browser/treeModel.ts:561]: I'm just replicating the behavior I found on the extensions, which also use a timeout to display the loading bar even after they finish loading. Also, I found out one of the listeners for one of the events here already removes the loading spinner upon completion of the refresh. If users won't see the refresh icon at all, would that be fine for them?

Comment (smartguest) [reply src/sql/base/parts/tree/browser/treeModel.ts:561]: I can remove the timeout part of the function, but users won't likely see anything if this is the case, for most ordinary things,

**Comment by smartguest** [comment]
I decided to set the ShowLoading attribute to true for the server tree, so that after a delay of 800 (which is the existing delay for it), it will show the spinner when the children are being refreshed. Once the children have been refreshed, the refreshed status is removed immediately. (This does not apply to when the item itself is refreshed). This mainly makes use of existing code, but just enables it to be used.

**Comment by smartguest** [comment]
Here's how the refresh logic works for the loading indicator in trees as I found in my findings:

There are two different types of trees: Async and Non Async, they both only show the loading bar/refresh icon if their children need to be refreshed.

For providers such as Azure, they utilize Async trees, and they call progressService on the dataProvider.getChildren call (without any delay provided), which tracks the getChildren promise until the call has returned at which point it hides the loading bar.

For Servers, it uses a regular tree, and it works differently:

you have to manually enable the showLoading attribute in the tree options, which is what my PR does.

here upon the start of the item's children refreshing, if the code finds the showLoading attribute being enabled, it will do a set timeout for 800 seconds (existing value not provided by me). The tree will then make a call to dataProvider.getChildren() as a promise, which upon completion, will fire an event called onDidRefreshChildren, which then triggers the removal of the loading status and the timeout.

So, in order for our code to replicate the exact experience for the server tree, we would need to remove the timeout or try to implement async tree for the servers (it is a completely different type) which could introduce more bugs and would require a ton more time to implement, more than the October deadline for this issue.

**Comment by Charles-Gagnon** [comment]
@smartguest Are you seeing this work for you on refresh? I pulled down your branch but I'm not seeing the spinner while it's refreshing a node

**Comment by smartguest** [comment]
It only briefly appears when the timeout is set to zero at first load. 

Literally the data provider loads so fast that it is impossible for the loading status to be set to true before the timeout is cancelled. If you want the spinner to be visible for the user, you have to add a timeout to it at the end.

**Comment by Charles-Gagnon** [comment]
What do you mean "it loads so fast"? Some of these queries to refresh the children take a long time so that definitely isn't always true. And this is the behavior I'm seeing - I click refresh and no loading spinner appears but 3+ seconds later the nodes are actually refreshed (with new nodes added or whatever). 

So basically this statement doesn't seem to be true (or isn't happening as I expect) - `which upon completion, will fire an event called onDidRefreshChildren, which then triggers the removal of the loading status and the timeout.`

Or do we do something weird like the "refresh" just kicks off the refresh and then returns, then when the children nodes are retrieved OE manually refreshes the tree itself? 

Regardless the original issue isn't being resolved as far as I can tell based on my tests mentioned above.

**Comment by Charles-Gagnon** [comment]
The Azure view is using a completely different tree structure (the async tree you mentioned) - so I'm not sure it's going to be easy to port over the logic from there into this older tree. But if you can find a way to do that then that'd be fine, although I'm curious what logic specifically you're talking about.

Switching OE to use the async tree is definitely what we should be aiming for in the long run, but that was something I was working on a long time back (the preview async tree stuff) and it ended up being a lot more difficult to do than anticipated so clearly is out of the scope of this PR.

**Comment by smartguest** [comment]
Timeout has been removed for more responsive refresh visibility, additionally the STS refresh call from hitting the action button has been moved to async server tree only (due to it being redundant for regular tree refresh). Loading status now shows up properly when object node takes time to refresh, no immediate issues detected.

**Comment by Charles-Gagnon** [comment]
Please provide an updated gif of this in action

File: src/sql/workbench/services/objectExplorer/browser/serverTreeDataSource.ts [hunk 7]
```diff
@@ -63,6 +63,8 @@ export class ServerTreeDataSource implements IDataSource {
 		} else if (element instanceof TreeNode) {
 			let node = element;
 			if (node.children) {
+				// Grab the latest data from the server of the node's children.
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/objectExplorer/browser/serverTreeDataSource.ts:66]: What happens if an error occurs during refresh? We already have logic to handle an error during the initial resolve - seems like that should probably apply here too.

File: src/sql/workbench/services/objectExplorer/browser/serverTreeDataSource.ts [hunk 8]
```diff
@@ -63,6 +63,8 @@ export class ServerTreeDataSource implements IDataSource {
...
+				await this._objectExplorerService.refreshTreeNode(node.getSession()!, node);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/objectExplorer/browser/serverTreeDataSource.ts:67]: @alanrenmsft I had Alex move this here, since the root problem we discovered is that it was calling refreshTreeNode and waiting for it to finish before it ever called tree.refresh - which meant that the actual tree refresh was also super quick (since it just immediately got the nodes) but the spinner never showed up since it wasn't told it was refreshing. 

I haven't dug myself into how getChildren is used though so I'm worried there might be perf side effects from this. Can you think of anything this might cause problems with? (and if you have time to dig through the code a bit to see if there's anything we missed that'd be great...)

**Comment by smartguest** [comment]
Gif has been added to main PR post.

File: src/sql/base/parts/tree/browser/treeView.ts [hunk 9]
```diff
@@ -392,7 +391,6 @@ function reactionEquals(one: _.IDragOverReaction, other: _.IDragOverReaction | n
 export class TreeView extends HeightMap {
 
 	static readonly BINDING = 'monaco-tree-row';
-	static readonly LOADING_DECORATION_DELAY = 800;
```
Comment (Charles-Gagnon) [inline src/sql/base/parts/tree/browser/treeView.ts:395]: Why did you remove this?

Comment (smartguest) [reply src/sql/base/parts/tree/browser/treeView.ts:395]: It was causing the refresh icon to not appear as often when there was this timeout. I removed the timeout so that the refresh would be more responsive to when the refresh action was taking place. If it is okay to only have the refresh icon appear occasionally (and to avoid brief flashes), I can restore it back to how it was before.

Comment (Charles-Gagnon) [reply src/sql/base/parts/tree/browser/treeView.ts:395]: I'm wondering if we should just reduce the delay instead. 800ms does seem like quite a long time to wait.

But I'm fine removing it completely for now - having it show even for a brief flash is pretty normal behavior for refresh actions in my experience.

Comment (smartguest) [reply src/sql/base/parts/tree/browser/treeView.ts:395]: I'll restore the timeout delay to about 100ms as a middle ground.

**Review by Charles-Gagnon — APPROVED** [review]
@kburtram Let's make sure that we're doing plenty of testing of this now and during release sign off.

**Review by cheenamalhotra — APPROVED** [review]

**Comment by cheenamalhotra** [comment]
Accidentally closed.. was just browsing through and noticed we're waiting @alanrenmsft to provide feedback.

Alan, would appreciate your feedback on this 🙂

File: src/sql/base/parts/tree/browser/treeView.ts [hunk 10]
```diff
@@ -392,7 +392,7 @@ function reactionEquals(one: _.IDragOverReaction, other: _.IDragOverReaction | n
...
+	static readonly LOADING_DECORATION_DELAY = 100;
```
Comment (alanrenmsft) [inline src/sql/base/parts/tree/browser/treeView.ts:395]: add a comment explaining what this is for and why 100 is chosen?

File: src/sql/workbench/services/objectExplorer/browser/serverTreeDataSource.ts (hunk 8 above)
Comment (alanrenmsft) [reply src/sql/workbench/services/objectExplorer/browser/serverTreeDataSource.ts:67]: let me dig into it a bit.

**Review by alanrenmsft — APPROVED** [review]
